source/particle/particle_handler.cc

```diff
@@ -222,6 +261,661 @@ namespace aspect
+             ExcInternalError());
+
+      // First clear the current ghost_particle information
+      ghost_particles.clear();
+
+      const std::map<types::subdomain_id, unsigned int> subdomain_to_neighbor_map(get_subdomain_id_to_neighbor_map());
+
+      std::vector<std::vector<particle_iterator> > ghost_particles_by_domain(subdomain_to_neighbor_map.size());
+
+      std::vector<std::set<unsigned int> > vertex_to_neighbor_subdomain(triangulation->n_vertices());
+
+      active_cell_it
+      cell = triangulation->begin_active(),
+      endc = triangulation->end();
+      for (; cell != endc; ++cell)
+        {
+          if (cell->is_ghost())
+            for (unsigned int v=0; v<GeometryInfo<dim>::vertices_per_cell; ++v)
+              vertex_to_neighbor_subdomain[cell->vertex_index(v)].insert(cell->subdomain_id());
+        }
+
+      cell = triangulation->begin_active();
+      for (; cell != endc; ++cell)
+        {
+          if (!cell->is_ghost())
+            {
+              std::set<unsigned int> cell_to_neighbor_subdomain;
+              for (unsigned int v=0; v<GeometryInfo<dim>::vertices_per_cell; ++v)
+                {
+                  cell_to_neighbor_subdomain.insert(vertex_to_neighbor_subdomain[cell->vertex_index(v)].begin(),
+                                                    vertex_to_neighbor_subdomain[cell->vertex_index(v)].end());
+                }
+
+              if (cell_to_neighbor_subdomain.size() > 0)
+                {
+                  const particle_range particles_in_cell = particle_range_in_cell(cell);
+
+                  for (std::set<types::subdomain_id>::const_iterator domain=cell_to_neighbor_subdomain.begin();
+                       domain != cell_to_neighbor_subdomain.end(); ++domain)
+                    {
+                      const unsigned int neighbor_id = subdomain_to_neighbor_map.find(*domain)->second;
+
+                      for (typename particle_range::iterator particle = particles_in_cell.begin(); particle != particles_in_cell.end(); ++particle)
+                        ghost_particles_by_domain[neighbor_id].push_back(particle);
+                    }
+                }
+            }
+        }
+
+      send_recv_particles(ghost_particles_by_domain,
+                          ghost_particles);
+    }
+
+
+
+    template <int dim, int spacedim>
+    void
+    ParticleHandler<dim,spacedim>::send_recv_particles(const std::vector<std::vector<particle_iterator> > &send_particles,
+                                                       std::multimap<types::LevelInd,Particle <dim> >     &received_particles,
+                                                       const std::vector<std::vector<active_cell_it> >    &send_cells)
+    {
+      // Determine the communication pattern
+      const std::vector<types::subdomain_id> neighbors (triangulation->ghost_owners().begin(),
+                                                        triangulation->ghost_owners().end());
```

Comment:
that's dangerous -- it only works if `ghost_owners()` returns a reference to whatever it is storing internally. If it is returning a *copy* of the object, then taking `begin` and `end` will yield pointers that point to separate objects.

It may be safer to copy what that function returns.